kuka_rsi_hw_interface/src/xml_handler/xml_element.cpp

```diff
@@ -0,0 +1,151 @@
+// Copyright 2023 Komáromi Sándor
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <map>
+#include <vector>
+#include <variant>
+#include <string>
+#include <cstring>
+#include <cstdlib>
+#include <iostream>
+#include <variant>
+
+#include "xml_handler/xml_element.hpp"
+
+namespace xml
+{
+bool XMLElement::CastParam(const XMLString & key, char * & str_ptr)
+{
+  auto param_it = params_.find(key);
+  auto ret_val = param_it != params_.end();
+  if (ret_val) {
+    errno = 0;
+    switch (param_it->second.param_type_) {
+      case XMLType::BOOL:
+        {
+          char * end = str_ptr;
+          auto res = (bool)strtol(str_ptr, &end, 0);
+          if (res == 0 && (errno != 0 || end == str_ptr)) {
+            return false;
+          }
+          str_ptr = end;
+          param_it->second.param_value_ = res;
+          break;
+        }
+      case XMLType::LONG:
+        {
+          char * end = str_ptr;
+          auto res = strtol(str_ptr, &end, 0);
+          if (res == 0 && (errno != 0 || end == str_ptr)) {
+            return false;
+          }
+          str_ptr = end;
+          param_it->second.param_value_ = res;
+          break;
+        }
+      case XMLType::DOUBLE:
+        {
+          char * end = str_ptr;
+          auto res = strtod(str_ptr, &end);
```

Comment:
IIRC, functions like `strtod(..)` use the current locale setting to decode floating point nrs ([doc](https://cplusplus.com/reference/cstdlib/strtod/)).

We've had problems in URDF parsing libraries on German and Dutch systems fi, where commas are used, versus `en_US` and similar locales where dots are used.

Reply:
Tanks for the tip, i should try out if RSI is locale-aware, if not I have to search for other solution

Reply:
You could enforce the C locale inside the driver/this class. That would always use dots AFAIK.

Reply:
The German documentation of RSI also uses '.'-s, so I would assume RSI is locale-independent

Reply:
Right. But what if the PC uses a German locale? IIUC (from the documentation), that would make `strtod(..)` expect `,`, not dots.

Reply:
I'm currently makeing the driver to use c locale in these functions, as you suggested earlier

Reply:
According to the setlocale [documentation](https://en.cppreference.com/w/cpp/locale/setlocale), the system always uses the C locale by default:
During program startup, the equivalent of std::setlocale([LC_ALL](http://en.cppreference.com/w/cpp/locale/LC_categories), "C"); is executed before any user code is run.


Therefore IMO it would not be needed to set the locale from the code

Reply:
I'm not saying you're wrong, but just wanted to link to the issues I mentioned earlier in the URDF parser. See https://github.com/ros/urdfdom_headers/issues/41 fi and the linked issues there.

Reply:
Is it possible that ROS itself set the locale to be the system locale? in our experiments we did not use ROS, just a simple C++ code

Reply:
I'm not aware of any code in ROS 2 which explicitly sets any locale.

(but then again, ROS is rather large, and I can't claim to have seen everything)